File: 0456-132-pattern/0456-132-pattern.java
class Solution {
    public boolean find132pattern(int[] arr) {
        int s=arr.length;
        Stack<Integer> temp=new Stack();
        int cur3=Integer.MIN_VALUE,cur1=Integer.MAX_VALUE;
        for(int i=s-1;i>=0;i--){
            cur1=arr[i];
            
            while(!temp.empty() && temp.peek()<arr[i]){
                cur3=temp.peek();
                temp.pop();
            }
            temp.push(arr[i]);
            if(cur1<cur3){
                return true;
            }
            
        }
        return false;
    }
}

File: 0456-132-pattern/0456-132-pattern.cpp
class Solution {
public:
    bool find132pattern(vector<int>& arr){
       int n = arr.size();
       int dpmini[n];
       dpmini[0] = arr[0];
       for(int i = 1; i<n; i++){
           dpmini[i] = min(arr[i], dpmini[i-1]);
       }
       stack<int> st;
       for(int i = n-1; i>=0; i--){
           int maxPopped = INT_MIN;
           while(!st.empty() && st.top() < arr[i]) {
               maxPopped = max(maxPopped, st.top());
               st.pop();
           }
           st.push(arr[i]);
           if(maxPopped > dpmini[i]) return true;
       }
       return false;
    }
};